doc/man3/EVP_PKEY_CTX_ctrl.pod

```diff
@@ -152,20 +254,58 @@ B<nid>. The B<nid> parameter must be B<NID_ffdhe2048>, B<NID_ffdhe3072>,
 B<NID_ffdhe4096>, B<NID_ffdhe6144> or B<NID_ffdhe8192>.  This macro can be
 called during parameter or key generation.
 
+The EVP_PKEY_CTX_set_dh_rfc5114() and EVP_PKEY_CTX_set_dhx_rfc5114() set the
+RFC5114 parameters instead of generating new DH parameters. Possible values are
+1, 2 and 3 corresponding to RFC5114 sections 2.1, 2.2 and 2.3. If set all other
+DH parameter options are overriden. 0 can be used to clear the stored value.
+
+The EVP_PKEY_CTX_set_dh_kdf_type() macro sets the key derivation function type
+to B<kdf> for DH key derivation. Possible values are B<EVP_PKEY_DH_KDF_NONE>
+and B<EVP_PKEY_DH_KDF_X9_42>.
+
+The EVP_PKEY_CTX_set0_dh_kdf_oid() macro sets the key derivation function
+object identifier to B<oid> for DH key derivation.
+
+The EVP_PKEY_CTX_set_dh_kdf_md() macro sets the key derivation function
+message digest to B<md> for DH key derivation.
+
+The EVP_PKEY_CTX_set_dh_kdf_outlen() macro sets the key derivation function
+output length to B<len> for DH key derivation.
+
+The EVP_PKEY_CTX_set0_dh_kdf_ukm() macro sets the user key material to
+B<ukm> for DH key derivation.
+
 The EVP_PKEY_CTX_set_ec_paramgen_curve_nid() sets the EC curve for EC parameter
 generation to B<nid>. For EC parameter generation this macro must be called
 or an error occurs because there is no default curve.
 This function can also be called to set the curve explicitly when
 generating an EC key.
 
-The EVP_PKEY_CTX_set_ec_param_enc() sets the EC parameter encoding to
+The EVP_PKEY_CTX_set_ec_param_enc() macro sets the EC parameter encoding to
 B<param_enc> when generating EC parameters or an EC key. The encoding can be
 B<OPENSSL_EC_EXPLICIT_CURVE> for explicit parameters (the default in versions
 of OpenSSL before 1.1.0) or B<OPENSSL_EC_NAMED_CURVE> to use named curve form.
 For maximum compatibility the named curve form should be used. Note: the
 B<OPENSSL_EC_NAMED_CURVE> value was only added to OpenSSL 1.1.0; previous
 versions should use 0 instead.
 
+The EVP_PKEY_CTX_set_ecdh_cofactor_mode() macro sets the cofactor mode to
+B<cofactor_mode> for ECDH key derivation. Possible values are 1 to enable
+cofactor generation, 0 to disable it and -1 to clear the stored cofactor key.
```

Comment:
Looking at the code in ec_pmeth it does look like doing a set(-1) does more than just clear the flag, it also frees memory (dctx->co_key) and defaults to the private ec key flags on subsequent get_cofactor_mode() calls. I am having trouble understanding what purpose this -1 parameter actually serves.

Reply:
The co_key is just a straight copy of the ec_key with the `EC_FLAG_COFACTOR_ECDH` flag set on it. Presumably so we don't want to make any modifications to the "real" key. We don't want to do anything in this EVP_PKEY_CTX that might affect future operations with this key (just because we want cofactor ECDH now, doesn't mean we will next time). AFAICT there is no real difference between passing 0 and -1 in terms of external behaviour. The former clears the flag on the co_key, and the latter frees the co_key. When it comes to the derivation step we use the co_key if it is present and otherwise use the ec_key, Using the co_key without the `EC_FLAG_COFACTOR_ECDH` flag set will get you exactly the same result as using the ec_key directly.